### PR TITLE
Cola layout option

### PR DESCRIPTION
**Pull request by YoonjinTKim**
Added cola layout into layout options. Click "#cola" will automatically layout the graph with original cola settings, and "#cola_setting" will direct to the #colaOptions HTML page to change the variables including randomizing, edge length, node space, fit, and flow options. The original setting is with {randomize:false, edge length = 45, node space = 5, fit = true, flow = undefined}. There are global variables in graphs_page.js to control and save the settings for cola options.

Added path to cola.js path to templates/base.html, templates/graphs/index.html, templates/graphs/index.html, reset.html, and templates/group/index.html.( <script src="https://bioinformatics.cs.vt.edu/~jeffl/cola.min.js"></script>, <script src="https://cdn.rawgit.com/cytoscape/cytoscape.js-cola/1.6.0/cytoscape-cola.js"></script>)

Create new navigation bar cola_option.html file with randomizing, edge length, node space, randomize, fit, and x -axis, and y-axis flows. 

In graph.js file, set global variables cola_edge_length, cola_node_space, cola_fit, cola_flow, and cola_randomize to store the settings. automatic_layout will call the initial settings for initial settings when the cola layout is first called.

image
<initial setting>

image
<Setting sidebar>

image
<y-axis flow>

### CODE REVIEW

**Comment by jlaw9** [comment]
#287

File: templates/base.html [hunk 1]
```diff
@@ -51,6 +51,15 @@
     <script src="{% static "components/lodash/dist/lodash.min.js" %}"></script>
     <script src="{% static "components/moment/min/moment.min.js" %}"></script>
     <script src="{% static "components/remarkable-bootstrap-notify/dist/bootstrap-notify.min.js" %}"></script>
+     <!-- <script src="https://code.jquery.com/jquery-1.11.3.min.js"></script> -->
+  <script src="http://js.cytoscape.org/js/cytoscape.min.js"></script>
+  <!-- <script src="https://github.com/tgdwyer/WebCola/blob/master/WebCola/cola.min.js"></script> -->
+  <!-- <script src="https://marvl.infotech.monash.edu/webcola/cola.v3.min.js"></script> -->
+  <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.1.1/jquery.min.js"></script>
```
Comment (adbharadwaj) [inline templates/base.html]: This line should go above all js libraries. It needs to be loaded first.

File: templates/base.html [hunk 2]
```diff
@@ -51,6 +51,15 @@
...
+  <!-- <script src="https://cdnjs.cloudflare.com/ajax/libs/cytoscape/2.7.16/cytoscape.min.js"></script> -->
+  <script src="https://bioinformatics.cs.vt.edu/~jeffl/cola.min.js"></script>
```
Comment (adbharadwaj) [inline templates/base.html]: Why are we loading from jeffs webserver. Can you load from bower package instead?

File: templates/base.html [hunk 3]
```diff
@@ -51,6 +51,15 @@
...
+  <script src="https://cdn.rawgit.com/cytoscape/cytoscape.js-cola/1.6.0/cytoscape-cola.js"></script>
```
Comment (adbharadwaj) [inline templates/base.html]: Again do not load js libraries or css from links...load them from bower components.

File: templates/base.html (hunk 2 above)
Comment (adbharadwaj) [inline templates/base.html]: This is actually not required in base.html.

File: templates/base.html (hunk 3 above)
Comment (adbharadwaj) [inline templates/base.html]: This is actually not required in base.html.

File: templates/graphs/reset.html [hunk 4]
```diff
@@ -44,6 +44,8 @@
 
     <script src="{% static 'graphs/js/jquery.min.js' %}"></script>
     <script src="{% static 'graphs/js/bootstrap.min.js' %}"></script>
+    <script src="https://bioinformatics.cs.vt.edu/~jeffl/cola.min.js"></script>
```
Comment (adbharadwaj) [inline templates/graphs/reset.html]: not required.

File: static/js/graphs_page.js [hunk 5]
```diff
@@ -1990,7 +1992,374 @@ var graphPage = {
             }
             graphPage.filterNodesEdges.showOnlyK();
         }
-    }
+    },
+
+
+    //--------------------------------------------------------------------------------------------
+    colaOption: {
```
Comment (adbharadwaj) [inline static/js/graphs_page.js]: Call it colaLayoutWidget

File: static/js/graphs_page.js [hunk 6]
```diff
@@ -1990,7 +1992,374 @@ var graphPage = {
...
+                flow:cola_flow,
+                infinite: cola_infinite,
+                maxSimulationTime: 1500
+            }
+
+            graphPage.applyLayout(newGraphLayout);
+            console.log("after applying layout");
+
+        },
+        overlapChanged : function (checked) {
```
Comment (adbharadwaj) [inline static/js/graphs_page.js]: All event handlers should start with "on". Like onOverlapChange. Make these changes to other event handlers as well in colaOption module

File: static/js/graphs_page.js [hunk 7]
```diff
@@ -1990,7 +1992,374 @@ var graphPage = {
...
+                fit: cola_fit,
+                flow: cola_flow,
+                infinite: cola_infinite,
+                maxSimulationTime: 1500
+            }
+
+            graphPage.applyLayout(newGraphLayout);
+            console.log("after applying layout");
+        },
+        randomizeChanged : function () {
```
Comment (adbharadwaj) [inline static/js/graphs_page.js]: Document what randomize is doing ?

File: static/js/graphs_page.js [hunk 8]
```diff
@@ -1990,7 +1992,374 @@ var graphPage = {
...
+                $("#input_node").val(50);
+
+            }
+            graphPage.colaOption.setBarToValueNodeSpace($("#input_node"), "slider2");
+            $("#slider2").slider({
+                value: $("#input_node").val(),
+                max: 50
+            });
+        },
+        applyMax: function (graph_layout) {
```
Comment (adbharadwaj) [inline static/js/graphs_page.js]: Function name is not descriptive enough. What is max ?If possible add link to cola documentation.

File: static/js/graphs_page.js [hunk 9]
```diff
@@ -2609,4 +2992,17 @@ var cytoscapeGraph = {
...
     }
 
-};
\ No newline at end of file
+};
+
+//variables for cola
+//Different from other layout options, cola can choose options for other variables
+//Initial edge length set up to 45 in range of 1 to 200
+var cola_edge_length = 45;
```
Comment (adbharadwaj) [inline static/js/graphs_page.js]: All of these variables need to go as parameter in colaOption or colaLayoutWidget module. Initialize them when the colaOption module's init function.

File: templates/base.html (hunk 2 above)
Comment (adbharadwaj) [inline templates/base.html]: not required on base html.

File: templates/base.html (hunk 3 above)
Comment (adbharadwaj) [inline templates/base.html]: not required.

File: templates/base.html [hunk 10]
```diff
@@ -51,6 +51,15 @@
     <script src="{% static "components/lodash/dist/lodash.min.js" %}"></script>
     <script src="{% static "components/moment/min/moment.min.js" %}"></script>
     <script src="{% static "components/remarkable-bootstrap-notify/dist/bootstrap-notify.min.js" %}"></script>
+     <!-- <script src="https://code.jquery.com/jquery-1.11.3.min.js"></script> -->
+  <script src="http://js.cytoscape.org/js/cytoscape.min.js"></script>
```
Comment (adbharadwaj) [inline templates/base.html]: not required.

File: templates/base.html (hunk 1 above)
Comment (adbharadwaj) [inline templates/base.html]: Move to top of all js libraries

File: static/js/graphs_page.js (hunk 9 above)
Comment (YoonjinTKim) [reply static/js/graphs_page.js]: I pushed the changes, and added cola.min.js file in static/components instead of calling Jeff's web url for cola.min.js

**Comment by adbharadwaj** [comment]
I am rejecting this pull request. I manually pulled in your code and made the changes I requested. You still have to submit a documentation. Please create a new pull request for documentation.

**Pull request merged**